### PR TITLE
routing methods type checking

### PR DESCRIPTION
**Pull request by RyanKung**
Do type checking when passing `methods`. 

I triggered a problem when tried passing a 'POST' as methods `prams` of routing function(actually flask.app.route), and the routing parse the string into ['P', 'O', 'S', 'T'], I think the best way is either ignore the wrong type or map it into the correct type, (to raise a exception is also a good choice) .

So there is the `pull request`: when passing a str type to `routing` as methods argument, try map it into a list type object : `str -> [str]`

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
this changes the external api, as such it needs a set of accompanying tests

File: werkzeug/routing.py [hunk 1]
```diff
@@ -613,6 +613,8 @@ def __init__(self, string, defaults=None, subdomain=None, methods=None,
         self.alias = alias
         if methods is None:
             self.methods = None
+        if isinstance(methods, str):
```
Comment (RonnyPfannschmidt) [inline werkzeug/routing.py]: this does not match the type and data correction done in the else branch

Comment (RyanKung) [inline werkzeug/routing.py]: sorry, and fixed

**Comment by RyanKung** [comment]
@RonnyPfannschmidt I added an unittest to cover that change

```
def test_pass_str_as_router_methods():
    test_get = r.Rule('/get', methods='GET')
    assert test_get.methods == set(['HEAD', 'GET'])
    test_post = r.Rule('/post', methods='POST')
    assert test_post.methods == set(['POST'])
```

**Comment by untitaker** [comment]
Would it make sense to explicitly reject strings instead? I think `methods="POST"` is both a common mistake and a hideous API.

**Comment by RyanKung** [comment]
@untitaker Yes, I knew the argument name `methods` is hinting that the `methods` should be a `IterableType`? So your suggestion is to raise an `TypeError` Exception?

**Comment by RyanKung** [comment]
```
if isinstance(methods, str):
    raise TypeError('methods must be IterableType, not `str`')
```

@untitaker I'm not a native Engilish speaker. I'm not sure that  `methods must be IterableType, not 'str'` or  `methods should not be str` which is better, or some other suggestion? Can you help me.

And I found that with `str` is also an `Iterable` type in this [definition](https://docs.python.org/3.5/glossary.html#term-iterable).

**Comment by untitaker** [comment]
I am not sure about the error message, but I think something along "you probably wanted to pass a single-element list".

On 7 May 2016 17:15:58 CEST, Ryan Kung <notifications@github.com> wrote:
>```
>if isinstance(methods, str):
>    raise TypeError('methods must be IterableType, not `str`')
>```
>
>@untitaker I'm not a native Engilish speaker. I'm not sure that 
>`methods must be IterableType, not 'str'` or  `methods should not be
>str` which is better, or some other suggestion? Can you help me.
>
>And I found that with `str` is also an `Iterable` type in this
>[definition](https://docs.python.org/3.5/glossary.html#term-iterable).
>
>
>
>
>
>---
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pallets/werkzeug/pull/927#issuecomment-217643588

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: tests/test_routing.py [hunk 2]
```diff
@@ -521,6 +521,14 @@ def test_implicit_head():
                   '/post', method='HEAD')
 
 
+def test_pass_str_as_router_methods():
+    try:
```
Comment (untitaker) [inline tests/test_routing.py]: It took me two reads to get what you were trying to achieve. Please use https://pytest.org/latest/assert.html#assertions-about-expected-exceptions

**Comment by RyanKung** [comment]
@untitaker Thanks and fixed.

**Comment by untitaker** [comment]
Excellent, thanks! The failing test is due to something else

**Pull request merged**